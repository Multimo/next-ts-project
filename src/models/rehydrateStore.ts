import { applySnapshot, getSnapshot } from 'mobx-state-tree';
import { alStore, albumStore, getAlbums, getPosts } from "./store";


export const rehydrateStore = (initialState?: any) => {
  initialState && applySnapshot(alStore, initialState);
  return alStore;
};


let store = null;
export const initStore = async (isServer: boolean) => {
  if (!isServer && alStore.albums.length && alStore.posts.length) {
    return albumStore;
  }
  const albums = await getAlbums();
  const posts = await getPosts();
  store = albumStore.create({
    albums,
    posts,
  });
  return {
    initialState: getSnapshot(store)
  };
};