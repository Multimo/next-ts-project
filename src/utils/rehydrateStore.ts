import { applySnapshot, getSnapshot } from 'mobx-state-tree';
import { alStore, albumStore } from "../models/store";


export const rehydrateStore = (initialState?: any) => {
  initialState && applySnapshot(alStore, initialState);
  return alStore;
};

interface InitialState {
  initialState: any;
}

let store = null;
export const initStore = (isServer: boolean, response: any,): InitialState | any => {
  if (!isServer) {
    applySnapshot(alStore, response)
    return albumStore;
  }

  store = albumStore.create(response);
  
  return {
    initialState: getSnapshot(store),
  };
};