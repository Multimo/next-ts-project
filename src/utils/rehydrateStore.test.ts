
import { rehydrateStore, initStore } from './rehydrateStore';

const mockPosts = {   
  posts: [{ 
    id: 123,
    userId: 123,
    title: 'ttile',
    body: 'body of the post goes here',
    comments: [],
  }] 
};

const mockBuiltStore = {
  "albums": [],
  "posts": [
    {
      "body": "body of the post goes here",
      "comments": [],
      "id": 123,
      "title": "ttile",
      "userId": 123,
    },
  ],
  "toggle": false,
};

describe('rehydrateStore tests' ,() => {
  it('initStore: should return initial state if used on server', () => {
      const expected = initStore(true, mockPosts);
      return expect(expected).toEqual({ initialState: mockBuiltStore })
  });

  it('rehydrateStore: should return store if initial state is falsy', () => {
    return expect(rehydrateStore(false)).toMatchSnapshot();
  });

  it('rehydrateStore: should return populated store if initial state', () => {
    return expect(rehydrateStore(mockPosts)).toEqual(mockBuiltStore);
  });
})
