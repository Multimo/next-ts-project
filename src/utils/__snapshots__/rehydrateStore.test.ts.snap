// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initStore: should return store if initial state is falsy 1`] = `
Object {
  "albums": Array [],
  "posts": Array [],
  "toggle": false,
}
`;

exports[`initStore: should return store if initial state is falsy 2`] = `
Object {
  "albums": Array [],
  "posts": Array [
    Object {
      "body": "body of the post goes here",
      "comments": Array [],
      "id": 123,
      "title": "ttile",
      "userId": 123,
    },
  ],
  "toggle": false,
}
`;

exports[`rehydrateStore tests rehydrateStore: should return store if initial state is falsy 1`] = `
Object {
  "albums": Array [],
  "posts": Array [],
  "toggle": false,
}
`;

exports[`rehydrateStore tests rehydrateStore: should return store if initial state is falsy 2`] = `
Object {
  "albums": Array [],
  "posts": Array [
    Object {
      "body": "body of the post goes here",
      "comments": Array [],
      "id": 123,
      "title": "ttile",
      "userId": 123,
    },
  ],
  "toggle": false,
}
`;
